### PR TITLE
Spanish Implementation

### PR DESCRIPTION
**Pull request by Laifsyn**
Submitting as draft because
1-) I'm unclear how pull requests actually works
2-) I haven't properly made an Integration Test because I found it extremely un-ergonomic to constantly build the Num2Words settings after every .to_word() call

I also submitted a .devcontainer configuration because I found it extremely helpful to be able to test stuffs in a virtual environment specially so when I didn't have available a beefy-enough computer (I might've forgotten to add Rust Docs Viewer in the extensions though)

### CODE REVIEW

**Comment by Ballasi** [comment]
Hey, sorry for the big delays in replying to the messages!

I'll take a look at your PR soon, thanks for the changes.

I'll be honest by saying I am not much interested in dev containers, do you think you can get them removed for now?

Also I've seen that you've done a lot of improvements regarding fmt, that's nice! I am wondering if it's too hard for you to split that in another PR? Usually one PR = one feature!

Thanks again :)

**Comment by Laifsyn** [comment]
oh yes. it's alright. 
As for the fmt it was something I ended up with after editing a template I copied and edited into my liking.

**Comment by Laifsyn** [comment]
just... give me a moment while I figure how to separate the stuffs out

**Comment by Laifsyn** [comment]
I'll assume the rustfmt thing will got to be in a separate PR right?

**Comment by Ballasi** [comment]
that would be best! just move files that are not changed in this PR to another one. If there was rustfmt ran within files that were changes within this PR, I am completely fine with keeping it here (on top of that, that'll avoid merge conflicts), makes it easier for both you and me :)

**Comment by Laifsyn** [comment]
Sheesh. It makes me glad I didn't go crazy with changing stuffs. fortunately I found out about how to revert a file to a specific commit. Hope this is the way to go for such situation

**Comment by Laifsyn** [comment]
Should I also undo the added Derives to the enums?

**Comment by Laifsyn** [comment]
Alright, this should be about it. (except for mod.rs because I was too lazy to undo rustfmt on that. Shame on me)
But everything should be the most minimal for the spanish implementation feature.

Rustfmt should be easily be doable in another pull request by just re-adding the `rustfmt.toml` file and running rustfmt against the project.

the undonde derives could be additionally done in another PR.

and optionally fix clippy warnings in another PR

**Review by Laifsyn — COMMENTED** [review]
I misread the commit content. This commit should've been `add missing docs`

**Comment by Ballasi** [comment]
Sorry for making you go through all of this trouble! and thank you very much for taking the time to work on it, it's very kind :)

I'll see the PR more in depth this weekend! If you're not planning on redoing the derives or rustfmt, I'll take on that job once this is merged.

**Comment by Laifsyn** [comment]
I was thinking about figuring what rustfmt config was the one that onelined the chained methods and see if it was alright to disable it.

The reason is because if the method chaining gets one-lined, then the type-hint for each method return on the chain will not be hinted by rust-analyser.

but anyways, the chainings tends to be short more often than not, so the **loss of the type-hint** which can only be seen in an IDE isn't that relevant

**Comment by Laifsyn** [comment]
Hi, I was thinking about making a [PR Benching with Criterion](https://github.com/Ballasi/num2words/pull/31) with another branch for benchmarks. Mainly  I want to see what issues can arises when I expect merge conflicts( expecting so because I had to add some derives to do the benchmark)

The reason was to see how the spanish implementation fared because I did a fix to the unneeded 0 spam that the split_thousands() method created. (Ref: [link](https://github.com/Ballasi/num2words/issues/11))

File: tests/lang_es_test.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
+use num2words::{Currency, Lang, Num2Err, Num2Words};
+use num_bigfloat::BigFloat;
+enum Outputs {
+    Cardinal,
+    Ordinal,
+    OrdinalNum,
+    Year,
+    Currency,
+}
+fn to_words(num: BigFloat, output: Outputs, preference: &[&str]) -> Result<String, Num2Err> {
```
Comment (Ballasi) [inline tests/lang_es_test.rs:10]: is there any reason why you are not using the actual builder each time, like on the other unit tests?

File: tests/lang_es_test.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+use num2words::{Currency, Lang, Num2Err, Num2Words};
```
Comment (Ballasi) [inline tests/lang_es_test.rs:1]: any reasons why you split the tests in half and have this here? imho it should be all under es.rs

File: README.md [hunk 3]
```diff
@@ -83,6 +83,7 @@ Here is a list of all of the supported languages:
 | 🇫🇷🇨🇦 | `Lang::French`    | `fr`      | French      | quarante-deux |
 | 🇧🇪🇨🇩 | `Lang::French_BE` | `fr_BE`   | French (BE) | quarante-deux |
 | 🇨🇭   | `Lang::French_CH` | `fr_CH`   | French (CH) | quarante-deux |
+| 🇪🇸   | `Lang::Spanish`   | `es`      | Spanish     | cuarenta y dos|
```
Comment (Ballasi) [inline README.md:86]: do put Spanish above French! I am thinking of ranking that per total number of speakers, and Spanish has more people speaking it than French

also thinking adding the flag of Mexico after the Spanish flag would make sense

File: src/bin/bin.rs [hunk 4]
```diff
@@ -25,6 +25,7 @@ AVAILABLE LANGUAGES:
     fr:      French (France and Canada)
     fr_BE:   French (Belgium and the Democratic Republic of the Congo)
     fr_CH:   French (Swiss Confederation and Aosta Valley)
+    es:      Spanish
```
Comment (Ballasi) [inline src/bin/bin.rs:28]: ditto

File: src/lang/es.rs [hunk 5]
```diff
@@ -0,0 +1,1044 @@
...
+            }
+
+            // Add the next Milliard if there's any.
+            if i != 0 && triplet != 0 {
+                if i > MILLARES.len() - 1 {
+                    return Err(Num2Err::CannotConvert);
+                }
+                // Boolean that checks if next Milliard is plural
+                let plural = triplet != 1;
+                match plural {
```
Comment (Ballasi) [inline src/lang/es.rs:534]: you should probably use a `if` case here rather than a `match` for bool (`if plural { ... } else { ... }`)

File: tests/lang_es_test.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    assert_eq!(word.unwrap(), "dos mil veintiuno");
+    let word = to_words(BigFloat::from(-2021), Outputs::Year, &[]);
+    assert_eq!(word.unwrap(), "dos mil veintiuno a. C.");
+
+    let word = to_words(BigFloat::from(21_001.21), Outputs::Currency, &[]);
+    assert_eq!(word.unwrap(), "veintiún mil un US dollars con veintiún centavos");
+
+    let word = to_words(BigFloat::from(21.01), Outputs::Currency, &[]);
+    assert_eq!(word.unwrap(), "veintiún US dollars con un centavo");
+}
```
Comment (Ballasi) [inline tests/lang_es_test.rs:102]: btw, compared the results from es.rs to python num2words and see differences

I don't mind much the differences in megas as it depends a lot on the source. however `mil trillones` makes me wonder, is this what should actually be the number?

```
10000000000000
diez billones
diez trillones

100000000000000
cien billones
cien trillones

1000000000000000000
un trillón
un quintillón

1000000000000000000000 <=== here
mil trillones
un sextillón

10000000000000000000000000
diez cuatrillones
diez septillones
```

Comment (Ballasi) [inline tests/lang_es_test.rs:102]: also see differences in ordinals that are worth looking at:

```
12
décimosegundo (python num2words)
duodécimo (your implementation)

1000
milésimo
primer milésimo
```

**Review by Ballasi — COMMENTED** [review]
Thank you again for taking the time to participate! I'm sorry I am relatively slow, I've got a lot going on lately.

Feel free to take a look at the comments I've put :)

File: tests/lang_es_test.rs (hunk 1 above)
Comment (Laifsyn) [reply tests/lang_es_test.rs:10]: two actually.
1-) I wanted to keep the `assert` line into 1 line if I could
2-) It was a bit easier to see what changed between one line and the other. (ngl, the Closure was overkill just to skip 1 parameter but welp...)

File: README.md (hunk 3 above)
Comment (Laifsyn) [reply README.md:86]: I originally put it below french because of alphabetic order. Curiously until now, popularity and alphabetic order matched lol
The mexican flag sounds alright. I just hope they don't have differing numbering (which remembers me, I need to see how to translate currency)

File: tests/lang_es_test.rs (hunk 2 above)
Comment (Laifsyn) [reply tests/lang_es_test.rs:1]: Edit: I misread

the `tests/*` folder is for integration tests

It simulates how the user would be using the library, and what output the user should expect.
Sort of similar to what the doc tests I wrote are for, but I felt too lazy to delete them after realising that they won't appear in the crate's docs

Tests inside `es.rs` would be to verify that internals (Which aren't public to outside the module) return the outputs that are expected. So if some shenanigans happens while linking stuffs to the public interface which the user will use, only the integration test can have possibility of breaking/failing. It legit is a waste of time, but I at the time (before figuring how integration tests sort of worked), I was more preoccupied with the internals instead of what the user would be getting

Disclaimer, I do however don't know if integration tests reflect in doc.rs. I believe they shouldn't

File: src/bin/bin.rs (hunk 4 above)
Comment (Laifsyn) [reply src/bin/bin.rs:28]: Ditto the pokemon?

File: src/lang/es.rs (hunk 5 above)
Comment (Laifsyn) [reply src/lang/es.rs:534]: I figured it looked slimmer with match instead of if
```rust
// preview:
if plural {
    words.push(String::from(MILLARES[i]))
} else {
    words.push(String::from(MILLAR[i]))
}

// vs match
match plural {
    false => words.push(String::from(MILLAR[i])),
    true => words.push(String::from(MILLARES[i])),
}
```

File: tests/lang_es_test.rs (hunk 6 above)
Comment (Laifsyn) [reply tests/lang_es_test.rs:102]: What's different?. Is Mil Trillones from Python?

I followed the standard of Each thousands changes the Millard because It was a bit confusing that a Billion*1000 is called the equivalent in spanish of "thousand billions" instead of "trillion"

Comment (Laifsyn) [reply tests/lang_es_test.rs:102]: https://es.wikipedia.org/wiki/Anexo:Nombres_de_los_n%C3%BAmeros_en_espa%C3%B1ol
So according to this source, it starts to jump by 1_000_000 instead of 1_000 after the trillion

As summary, English's [Sextillion](https://en.wikipedia.org/wiki/English_numerals#Very_large_numbers) is 10^21
Spanish's Sextillion (according to [this particular page in wiki](https://es.wikipedia.org/wiki/Sextill%C3%B3n)) is 10^36

I guess it would still best to discuss with someone more familiarized with spanish, because if this really is the spanish standard, I'm pretty sure almost none will get spanish's quadrillon right (Because it actually is "A Million Trillion")

Comment (Laifsyn) [reply tests/lang_es_test.rs:102]: "décimosegundo" could be just pyton's laziness to this edge case. But they're actually the same. From the Spanish Language academy: [duodécimo](https://dle.rae.es/duod%C3%A9cimo)

as for the 1000, I was really second guessing if ordinals also had the same rules as Cardinal when it regarded to the thousands. I'll probably add the edge case

Comment (Laifsyn) [reply tests/lang_es_test.rs:102]: also regarding the differences between `duodécimo` and `décimosegundo`, from quoted from https://www.rae.es/dpd/ordinales
`
c) Los ordinales correspondientes a los números 11 y 12 presentan hoy dos formas: las etimológicas simples undécimo y duodécimo, que son las preferidas en el uso culto, y las compuestas decimoprimero y decimosegundo, creadas modernamente por analogía con la forma que adoptan los ordinales del resto de las series (vigesimoprimero, vigesimosegundo, trigésimo primero, trigésimo segundo, etc.).
`
So as  a summary, the biword terms of `decimoprimero` and `decimosegundo` are just terms that has been come up with as an analogy to the rest of the ordinals expression

**Comment by Laifsyn** [comment]
Alright. That should be it

**Comment by Laifsyn** [comment]
Oh, btw. I'm working on translating the currencies. Might submit a commit by the end of the day

**Comment by Laifsyn** [comment]
I have a question if you can get the answer

I'll be leaving it as the first meanwhile, but should it be like the second case?

```rust
assert_eq!(
    ordinal(124_001_091_000_000_000_001),
    "ciento veinticuatrotrillonésimos milnoventa y unobillonésimos primeros"
//  "ciento veinticuatrotrillonésimos mil noventa y unobillonésimos primeros"
);
```

**Comment by Laifsyn** [comment]
I couldn't redo the Cardinal implementation.
Performance gains were dirt low. at about x1.02-x1.25 faster, and also the implementation looked a bit ugly. So I have no changes to the current implementation of Spanish.

**Comment by PabloGmz96** [comment]
Hey guys! Native Spanish speaker here. Is there anything that you need help with? I'd love to use this crate in a project that I'm currently developing

**Comment by Laifsyn** [comment]
> Hey guys! Native Spanish speaker here. Is there anything that you need help with? I'd love to use this crate in a project that I'm currently developing
- [ ] I would like opinions on whether to support a flavor for "Semantically Incorrect" milliards. 
according to  RAE, 10E9(1,000,000,000) is called "Billón", whereas in English, 10E9 is called "Trillion". Choosing to support semantically Incorrect cardinals would mean that ~~spanish's "Billón" would be english's trillion~~ milliards like "billón" and "trillón" in spanish are just the literal translation of english's.

- [ ] Maybe check currency, in case some currency got mistranslated.


- [ ] ~~Support converting too big of numbers into `n`E`m` expression, i.e. `20.23E20` => `veinte coma veintitrés por 10 a la 20`~~

Priority close to 0:
* Cleanup tests & Integration Tests: Because some test cases overlaps other tests. Also I personally think they could look cleaner with some work.

Otherwise, everything should be fine. Other than big numbers which can lose precision

**Comment by PabloGmz96** [comment]
Hey @Laifsyn! I just review the "Semantically Incorrect" numbers that you mention and as far as I know almost the complete Latin-america region use those big numbers exactly in the same way as in English, so a direct translation of them should be precise enough. Later versions of this crate could include some other forms of Spanish cardinals numbers but knowing that they are rarely used this should be not a priority. But let me ask some friends in Colombia and Argentina (Countries that use particularly high number prices in the every day life) to confirm that.

In the other hand, let me test the USD and PESOS translation, those are the currencies that I'm familiar with.

And last but not least, I just want to say a big thank you to make the Spanish implementation possible!

**Comment by PabloGmz96** [comment]
Just a quick question... I'm using this crate in library mode, is there any way to just print the resulting words? The output always show me the array that the code use to convert numbers to words.
image

**Comment by PabloGmz96** [comment]
Just a quick update... When using Spanish as language, the output shows the array that the code use to translate the numbers to words.
image

**Comment by Laifsyn** [comment]
> Just a quick update... When using Spanish as language, the output shows the array that the code use to translate the numbers to words. image

Thanks for finding out! I might have forgotten to test the actual binary before I uploaded

**Comment by bd-g** [comment]
Any chance on getting this merged and pushed out?

**Comment by PabloGmz96** [comment]
I just want to add that a year after PROD usage of this branch, I've found zero issues in Cardinal translation!

**Comment by Laifsyn** [comment]
> I just want to add that a year after PROD usage of this branch, I've found zero issues in Cardinal translation!

Wait, so it means there were issues in other places? 💀
